jax_triton/pallas/pallas_call.py

```diff
@@ -221,42 +269,110 @@ def clear_caches():
   _initial_style_open_jaxpr.cache_clear()
   _compile_jaxpr.cache_clear()
 
-def pallas_call(f, out_shape, grid, *, debug: bool = False,
+def _preprocess_grid(grid: Optional[Union[Grid, int]]) -> Grid:
+  if grid is None:
+    return (1,)
+  if isinstance(grid, int):
+    return (grid,)
+  return grid
+
+def _extract_function_name(f: Callable, name: Optional[str]) -> str:
+  if name is None:
+    name = f.__name__ if hasattr(f, "__name__") and f.__name__ else "func"
+  return name
+
+def _convert_block_spec_to_block_mapping(
+    grid: Grid, block_spec: Optional[BlockSpec]) -> Optional[BlockMapping]:
+  if block_spec is None:
+    return None
+  in_avals = [jax_core.ShapedArray((), jnp.int32) for _ in grid]
+  jaxpr, _, consts = pe.trace_to_jaxpr_dynamic(
+      lu.wrap_init(block_spec.index_map), in_avals)
+  if consts:
+    raise NotImplementedError
+  return BlockMapping(block_spec.block_shape, jax_core.ClosedJaxpr(jaxpr, consts))
+
+def pallas_call(f: Callable, out_shape: Any, *, debug: bool = False,
+                grid: Optional[Grid] = None,
+                in_specs: Optional[Sequence[Optional[BlockSpec]]] = None,
+                out_specs: Optional[Sequence[Optional[BlockSpec]]] = None,
                 num_warps: int = 4, num_stages: int = 3,
                 input_output_aliases: Dict[int, int] = {},
                 interpret: bool = False,
-                name: Optional[str] = None,
-                **metaparams):
-  if isinstance(grid, int):
-    grid = (grid,)
-  if name is not None:
-    name = name
-  else:
-    name = f.__name__ if hasattr(f, "__name__") and f.__name__ else "func"
+                name: Optional[str] = None):
+  grid = _preprocess_grid(grid)
+  name = _extract_function_name(f, name)
+  singleton = False
+  if not isinstance(out_shape, (tuple, list)):
+    out_shape = (out_shape,)
+    singleton = True
   flat_out_shapes, out_tree = tree_util.tree_flatten(out_shape)
-  flat_out_shapes = [jax.ShapeDtypeStruct(x.shape, x.dtype) for x in
-                     flat_out_shapes]
-  def wrapped(*args, **kwargs):
-    flat_args, _ = tree_util.tree_flatten((args, kwargs))
-    in_tree = tree_util.tree_structure((*args, *flat_out_shapes))
-    flat_avals = map(
-        jax_core.raise_to_shaped, map(jax_core.get_aval, flat_args))
-    flat_output_avals = [jax_core.ShapedArray(out_shape.shape, out_shape.dtype)
-        for out_shape in flat_out_shapes]
-    ptr_avals = [state.ShapedArrayRef(aval.shape, aval.dtype) for aval in flat_avals]
-    out_ptr_avals = [state.ShapedArrayRef(aval.shape, aval.dtype) for aval in flat_output_avals]
+  if out_specs is not None:
+    out_specs = tree_util.tree_unflatten(
+        out_tree, tree_util.tree_leaves(out_specs))
```

Comment:
I have to say this is a bit weird to ignore the structure of `out_specs`. It would be better to follow the convention all the `*map` functions use and do pytree prefix broadcasting

Reply:
Also, is that variable even used? You recompute `out_specs` below anyway

Reply:
> I have to say this is a bit weird to ignore the structure of out_specs. It would be better to follow the convention all the *map functions use and do pytree prefix broadcasting

Yea I think this is the right thing to do in the long run. I modified it to just convert both to tuples.